src/components/ban-sync.ts

```diff
@@ -43,19 +49,106 @@ const supportedRecommendations = [
     "org.matrix.mjolnir.ban", // Used historically.
     "m.ban"
 ];
+
+interface MatrixRegistrationResponse {
+    flows: {
+        stages: string[],
+    }[],
+}
+
+enum RegistrationStatus {
+    Unknown = 0,
+    Open = 1,
+    ProtectedEmail = 2,
+    ProtectedCaptcha = 3,
+    Closed = 4,
+}
+
+const AuthTypeRecaptcha = 'm.login.recaptcha';
+const AuthTypeEmail = 'm.login.email.identity';
+
 /**
  * Synchronises Matrix `m.policy.rule` events with the bridge to filter specific
  * users from using the service.
  */
 export class MatrixBanSync {
-    private bannedEntites = new Map<string, BanEntity>();
-    private subscribedRooms = new Set<string>();
-    constructor(private config: MatrixBanSyncConfig) { }
+    private readonly homeserverPropertiesCache = new Map<string, {openRegistration: RegistrationStatus; ts: number;}>();
+    private readonly bannedEntites = new Map<string, BanEntity>();
+    private readonly subscribedRooms = new Set<string>();
+    private readonly hostResolver = new MatrixHostResolver();
+    constructor(private config: MatrixBanSyncConfig) {
+
+    }
+
+    public async getHomeserverProperties(serverName: string) {
+        const hsData = this.homeserverPropertiesCache.get(serverName);
+        // Slightly fuzz the ttl.
+        const ttl = CACHE_HOMESERVER_PROPERTIES_FOR_MS + (Math.random()*60000);
+        if (hsData && hsData.ts < ttl) {
+            return hsData;
+        }
+
+        const { url } = await this.hostResolver.resolveMatrixServer(serverName);
+        const registrationResponse = await axios.post(new URL('/_matrix/client/v3/register', url).toString(), { }, { });
```

Comment:
What about older homeservers? If old homeservers only expose `/_matrix/client/r0/register`, will the bridge allow them to stay despite having open reg?

Reply:
Fair point, well include those.